plugins/chain_plugin/trx_retry_db.cpp

```diff
@@ -0,0 +1,326 @@
+         } );
+      }
+   }
+
+   void on_block_start( uint32_t block_num ) {
+      // on forks rollback any accepted block transactions
+      rollback_to( block_num );
+   }
+
+   void on_accepted_block(const chain::block_state_ptr& bsp ) {
+      // good time to perform processing
+      ack_ready_trxs_by_block_num( bsp->block_num );
+      retry_trxs();
+   }
+
+   void on_irreversible_block(const chain::block_state_ptr& bsp ) {
+      ack_ready_trxs_by_lib( bsp->block_num );
+      clear_expired( bsp->block->timestamp );
+   }
+
+private:
+
+   void rollback_to( uint32_t block_num ) {
+      const auto& idx = _tracked_trxs.index().get<by_block_num>();
+      // determine what to rollback
+      std::vector<decltype(_tracked_trxs.index().project<0>(idx.begin()))> to_process;
+      for( auto i = idx.rbegin(); i != idx.rend(); ++i ) {
+         // called on block_start, so any block_num greater or equal have been rolled back
+         if( i->block_num < block_num ) break;
+
+         auto ii = i.base(); // switch to forward iterator, then -- to get back to item
+         to_process.emplace_back( _tracked_trxs.index().project<0>( --ii ) );
+      }
+      // perform rollback
+      for( auto& i : to_process ) {
+         _tracked_trxs.modify( i, [&]( tracked_transaction& tt ) {
+            // if forked out, then need to retry. Estimate last_try via block_num it was in.
+            tt.last_try = fc::time_point::now() - fc::microseconds( (i->block_num - tt.block_num) * chain::config::block_interval_us );
+            tt.block_num = 0;
+            tt.trx_trace_v.clear();
+         } );
+      }
+   }
+
+   void retry_trxs() {
+      const auto& idx = _tracked_trxs.index().get<by_last_try>();
+      auto now = fc::time_point::now();
+      // determine what to retry
+      std::vector<decltype(_tracked_trxs.index().project<0>(idx.begin()))> to_process;
+      for( auto i = idx.begin(); i != idx.end(); ++i ) {
+         if( i->is_ready() ) break;
+
+         if( i->last_try + _retry_interval <= now ) {
+            to_process.emplace_back( _tracked_trxs.index().project<0>( i ) );
+         }
+      }
+      // retry
+      for( auto& i: to_process ) {
+         _transaction_ack_channel.publish(
+               appbase::priority::low, std::pair<fc::exception_ptr, packed_transaction_ptr>( nullptr, i->ptrx ) );
+         _tracked_trxs.modify( i, [&]( tracked_transaction& tt ) {
+            tt.last_try = now;
+         } );
+      }
+   }
+
+   void ack_ready_trxs_by_block_num( uint32_t block_num ) {
+      const auto& idx = _tracked_trxs.index().get<by_ready_block_num>(); // less sort
+      // determine what to ack
+      std::vector<decltype(idx.begin())> to_process;
+      for( auto i = idx.begin(); i != idx.end(); ++i ) {
+         if( !i->is_ready() ) break;
+         if( i->num_blocks == lib_totem ) break;
+         // if we have reached requested block height then ack to user
+         if( i->ready_block_num() <= block_num ) {
```

Comment:
Couldn't we move this to the predicate for the for statement, i.e. can't we quick once we hit an iterator with ready_block_num() higher than block_num? And if that is correct, we can get rid of lines 224 and 225 as well.

Reply:
Yes. And can use upper_bound like you suggested below.